---
title: "Lab_10"
output: html_document
author: Alina Sequeira
date: 11-5-20
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r}
library(vcfR)
```

## VCF File Structure
```{r}
data(vcfR_example)
vcf
```

## The Meta Region
```{r}
strwrap(vcf@meta[1:7])
```
```{r}
# use this function to make metadata easier to read
queryMETA(vcf)
```
```{r}
# returns info about element DP
queryMETA(vcf, element = 'DP')
```
```{r}
# more specific element parameter
queryMETA(vcf, element = 'FORMAT=<ID=DP')
```

## The Fix Region
The fix region contains information for each variant which is sometimes summarized over all samples.
The first eight columns of the fixed region are titled CHROM, POS, ID, REF, ALT, QUAL, FILTER and INFO.
The function getFIX() will suppress the INFO column by default
```{r}
head(getFIX(vcf))
```

## The gt Region
The gt (genotype) region contains information about each variant for each sample
The values for each variant and each sample are colon delimited - can store multiple types of genotypes
The format of the data is specified by the FORMAT column (column nine)
```{r}
vcf@gt[1:6, 1:4]
```

## vcFR
```{r}
head(vcf)
```
```{r}
write.vcf(vcf, "data/myVCFdata_filtered.vcf.gz")
```

## Exercises Part 1
How would you find more information about read.vcfR()?
```{r}
?read.vcfR
```

How would we learn what the acronym “AD” stands for?
```{r}
# use queryMETA() @ specific element 

queryMETA(vcf, element = 'AD')
```

We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?
```{r}
# use tail() 

tail(vcf@fix)
```

There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?
```{r}
plot(vcf)
```
How would we query the sample names?
```{r}
colnames(vcf@gt)
```

## Part 2 Analysis of Genome Data
The show method reports that we have 18 samples and 2,190 variants. If this matches our expectation then we can proceed.
```{r}
vcf
```

Use the function vcfR2genlight() to convert our vcfR object to a genlight object.
This makes our VCF data available to the analyses in adegenet.
```{r}
library(adegenet)
```
```{r}
x <- vcfR2genlight(vcf)
```
```{r}
x
```
```{r}
# vcfR
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]
```
```{r}
# genlight
t(as.matrix(x))[c(1,5,17), 1:3]
```
```{r}
pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)
```
```{r}
ploidy(x) <- 2
```

## Distance Matrices
create a pairwise genetic distance matrix for individuals or populations (i.e., groups of individuals)
create distance matrix from a genlight object using dist()
```{r}
x.dist <- dist(x)
```
also can use bitwise.dist() function to create distance matrices from genlight objs
```{r}
library(poppr)
```
```{r}
x.dist <- poppr::bitwise.dist(x)
```

## chromR Objects
using chromR to locate unusual features in a genome
helps visualize the effects of manipulations

## Creating chromR 
Warning messages indicate that the names in all the data sources do not perfectly match
```{r}
library(vcfR)

# Find the files.
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```
Once we have created our chromR object we can verify that its contents are what we expect
The show method for chromR objects presents a summary of the object’s contents.
```{r}
chrom
```
There at least two ways to graphically view the chromR object. The first is plot() which plots histograms of some of data summaries.
```{r}
# Qual is hard to interpret - most variants of low quality - QUAL not a good parameter for this data to judge variants
# SNP densities is empty b/c the data is creating during the process of chromR objects
plot(chrom)
```
second plot, called chromo plot, displays the same information as the plot method only it distributes the data along its chromosomal coordinates. *includes a representation of the annotation data
```{r}
chromoqc(chrom, dp.alpha = 66)
```

## Processing chromR Objects
```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
```
```{r}
plot(chrom)
```
When we observe the chromoqc plot we see that we now have variant densities, nucleotide content as well as a representation of where in our reference we have nucleotides (A, C, G or T) or where we have ambiguous nucleotides.
```{r}
#  an example of visualizing raw data that has come from a variant caller and other automated sources
chromoqc(chrom, dp.alpha = 66)
```
We can use this data to create a chromR object and compare it to the above (raw) data.
```{r}
#vcf <- read.vcfR("data/myVCFdata_filtered.vcf.gz", verbose = FALSE)

vcf <- read.vcfR("data/myVCFdata_filtered.vcf.gz", verbose = FALSE)
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```
```{r}
# more uniform read depth
# consistent mapping quality
```

## Tabular Summaries
When we process a chromR object, two forms of tabular data are created
summaries are made on a per variant basis (sample size, allele counts, heterozygosity, and effective size)
summaries made on a per window basis
include nucleotide content per window (including missing data so you can adjust window size for analyses if necessary), the number of genic sites per window (when annotation information was provided) and the number of variants per window.
window size can be changed w/ win.size
```{r}
head(chrom@var.info)
```
```{r}
head(chrom@win.info)
```

## Genetic Differentiation
function genetic_diff() - measure population diversity and differentiation
```{r}
library(vcfR)
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
```
```{r}
knitr::kable(head(myDiff[,16:19]))
```
summarize, take averages of data
```{r}
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```
summarize, use violin plots
```{r}
library(reshape2)
library(ggplot2)

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)
```
```{r}
p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

## Exercises Part II
You actually have everything you need to make a Manhattan plot. Can you figure out how to plot G′ST (y-axis) by genomic position (POS)?
```{r}
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
```
This Manhatttan plot shouldlook a bit unusual. Can you think of anything that may be wrong with this analysis?
```{r}
# the sample size is too small

table(pop)
```
Can you figure out how to zoom in on a particular region of a chromosome in chromoqc()?
```{r}
chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```
Can you use the function queryMETA() to look for other data in your file that may be of interest?
```{r}
queryMETA(vcf)
```

